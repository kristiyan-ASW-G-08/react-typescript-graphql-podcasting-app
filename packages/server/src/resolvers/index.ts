import userMutations from 'src/users/usersMutations';
import podcastMutations from '@src/podcasts/podcastsMutations';
import podcastQueries from '@src/podcasts/podcastQueries';
import episodeMutations from '@src/episodes/episodeMutations';

const resolvers = {
  Query: {
    ...podcastQueries,
  },
  Mutation: { ...userMutations, ...podcastMutations, ...episodeMutations },
};

export default resolvers;
