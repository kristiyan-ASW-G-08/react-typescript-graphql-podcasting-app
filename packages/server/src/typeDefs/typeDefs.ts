import { gql } from 'apollo-server';

const typeDefs = gql`
  scalar Upload

  type User {
    username: String!
    email: String!
    _id: ID!
  }

  type Query {
    getUser: User!
  }

  type File {
    filename: String!
    mimetype: String!
    encoding: String!
  }
  type UserLoginPayload {
    username: String!
    email: String!
    userId: String!
    token: String!
  }
  type Podcast {
    title: String!
    website: String!
    cover: String!
    user: User!
    _id: ID!
  }

  type Episode {
    title: String!
    description: String!
    audioFile: String!
    date: String!
    user: User!
    podcast: Podcast!
    _id: ID!
  }
  type Episodes {
    episodes: [Episode]!
  }
  type Query {
    podcasts(query: String): [Podcast]!
    getPodcast(podcastId: ID!): Podcast!
    getEpisode(episodeId: ID!): Episode!
    getEpisodesByPodcast(podcastId: ID!): Episodes!
  }
  type Mutation {
    createUserMutation(
      username: String!
      email: String!
      password: String!
      confirmPassword: String!
    ): UserLoginPayload!

    loginMutation(email: String!, password: String!): UserLoginPayload!

    createPodcastMutation(
      title: String!
      website: String!
      cover: Upload!
    ): Podcast!

    updatePodcastMutation(
      title: String!
      website: String!
      cover: Upload
      _id: ID!
    ): Podcast!

    createEpisodeMutation(
      title: String!
      description: String!
      audioFile: Upload!
      podcast: ID!
    ): Episode!
  }
`;

export default typeDefs;
