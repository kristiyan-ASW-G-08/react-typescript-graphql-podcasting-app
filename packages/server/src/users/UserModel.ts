import mongoose, { Schema } from 'mongoose';
import UserType from '@src/types/UserType';
import duplicationErrorHandler from '@customUtilities/duplicationErrorHandler';
import uniqueValidator from 'mongoose-unique-validator';

const UserSchema: Schema = new Schema({
  username: { type: String, minlength: 1, maxlength: 50, unique: true },
  email: { type: String, unique: true },
  password: { type: String, minlength: 8 },
});

UserSchema.plugin(uniqueValidator);
// @ts-ignore
UserSchema.post('save', duplicationErrorHandler);
// @ts-ignore
UserSchema.post('update', duplicationErrorHandler);

export default mongoose.model<UserType>('User', UserSchema);
