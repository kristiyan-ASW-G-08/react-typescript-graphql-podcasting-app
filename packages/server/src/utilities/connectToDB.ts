import mongoose from 'mongoose';
import logger from '@customUtilities/logger';

const connectToDB = async (mongoURI: string): Promise<void> => {
  try {
    await mongoose.connect(mongoURI, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
  } catch (err) {
    logger.error(`MongoDB Connection Error: ${err}`);
  }
};
export default connectToDB;
