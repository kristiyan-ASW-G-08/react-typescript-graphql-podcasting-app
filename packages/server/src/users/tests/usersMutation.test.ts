import connectToDB from '@customUtilities/connectToDB';
import mongoose from 'mongoose';
import UserModel from '@src/users/UserModel';
// @ts-ignore
import typeDefs from '@typeDefs/typeDefs';
import apolloServer from '@customUtilities/createApoloServer';
import { string } from 'yup';
import { ApolloServer } from 'apollo-server';
import bcrypt from 'bcryptjs';
import resolvers from '../../resolvers/index';

describe('Resolvers', () => {
  const { MONGO_USER, MONGO_PASSWORD, MONGO_DATABASE } = process.env;
  const mongoURI = `mongodb+srv://${MONGO_USER}:${MONGO_PASSWORD}@cluster0.ol9wi.mongodb.net/${MONGO_DATABASE}?retryWrites=true`;
  beforeAll(async () => {
    await mongoose.disconnect();
    await connectToDB(mongoURI);
    await UserModel.deleteMany({}).exec();
  });

  afterEach(async () => {
    await UserModel.deleteMany({}).exec();
  });
  afterAll(async () => {
    await mongoose.disconnect();
  });

  describe('Mutations', () => {
    afterEach(async () => {
      await UserModel.deleteMany({}).exec();
    });
    const values = {
      username: 'username',
      email: 'somemail@mail.test',
      password: 'correctPassword',
      confirmPAssword: 'correctPassword',
    };

    const incorrectValues = {
      username: '',
      email: 'notAnEMail',
      password: 'pass',
      confirmPAssword: 'correctPassword',
    };
    describe('loginMutation', () => {
      it('login successfully', async () => {
        expect.assertions(2);

        await new UserModel({
          username: values.username,
          email: values.email,
          password: await bcrypt.hash(values.password, 12),
        }).save();
        const query = `mutation {
          loginMutation(email:"${values.email}",password:"${values.password}",) {
            token
          }
        }`;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
        });
        expect(result.errors).toBeUndefined();
        expect(result.data?.loginMutation.token).toBeTruthy();
      });

      it('login unsuccessfully', async () => {
        expect.assertions(1);

        await new UserModel({
          username: values.username,
          email: values.email,
          password: await bcrypt.hash(values.password, 12),
        }).save();
        const query = `mutation {
          loginMutation(email:"${values.email}",password:"${incorrectValues.password}",) {
            token
          }
        }`;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
        });
        expect(result.errors).toMatchSnapshot();
      });
      it('login unsuccessfully because user is not found', async () => {
        expect.assertions(1);
        const query = `mutation {
          loginMutation(email:"${values.email}",password:"${incorrectValues.password}",) {
            token
          }
        }`;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
        });
        expect(result.errors).toMatchSnapshot();
      });
    });
    describe('createUserMutation', () => {
      it('it creates and returns a User successfully', async () => {
        expect.assertions(2);
        const query = `mutation {
          createUserMutation(username:"${values.username}",email:"${values.email}",password:"${values.password}",confirmPassword:"${values.confirmPAssword}"){
            id
          }
        }`;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
        });

        expect(result.errors).toBeUndefined();
        expect(result.data?.createUserMutation.id).toBeTruthy();
      });
      it('it fails validation and returns validation errors', async () => {
        expect.assertions(1);
        const query = `mutation {
          createUserMutation(username:"${incorrectValues.username}",email:"${incorrectValues.email}",password:"${incorrectValues.password}",confirmPassword:"${incorrectValues.confirmPAssword}"){
            id
          }
        }`;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
        });

        // @ts-ignore
        expect(result.errors[0].extensions?.validationErrors).toMatchSnapshot();
      });

      it('it uses duplicate email and username and returns validation errors', async () => {
        expect.assertions(1);

        await new UserModel({
          username: values.username,
          email: values.email,
          password: values.password,
        }).save();
        const query = `mutation {
          createUserMutation(username:"${values.username}",email:"${values.email}",password:"${values.password}",confirmPassword:"${values.confirmPAssword}"){
            id
          }
        }`;

        const testServer = new ApolloServer({
          typeDefs,
          resolvers,
        });

        const result = await testServer.executeOperation({
          query,
        });
        // @ts-ignore
        expect(result.errors[0].extensions?.validationErrors).toMatchSnapshot();
      });
    });
  });
});
