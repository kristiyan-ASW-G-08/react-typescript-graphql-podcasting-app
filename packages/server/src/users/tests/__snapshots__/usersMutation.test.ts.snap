// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers Mutations createUserMutation it fails validation and returns validation errors 1`] = `undefined`;

exports[`Resolvers Mutations createUserMutation it uses duplicate email and username and returns validation errors 1`] = `undefined`;

exports[`Resolvers Mutations loginMutation login unsuccessfully 1`] = `
Array [
  [GraphQLError: Validation Error],
]
`;

exports[`Resolvers Mutations loginMutation login unsuccessfully because user is not found 1`] = `
Array [
  [GraphQLError: Validation Error],
]
`;
